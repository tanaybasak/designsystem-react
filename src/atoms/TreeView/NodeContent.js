import React from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
const NodeContent = ({ content }) => {
  return (
    <div
      className={`${prefix}-tree-content-node`}
      style={{ flexBasis: 'calc(100% - 50px)' }}
      title={content}
    >
      <div className="tree-content-text">{content}</div>
    </div>
  );
};

NodeContent.propTypes = {
  content: PropTypes.any
};

NodeContent.defaultProps = {
  content: null
};

export default NodeContent;
