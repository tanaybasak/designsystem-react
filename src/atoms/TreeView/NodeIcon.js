import React, { cloneElement, useContext } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';

import TreeStateContext from './treeStateContext';
import TreeFunctionContext from './treeFunctionContext';

const NodeIcon = ({ node }) => {
  const state = useContext(TreeStateContext);
  const callbackContext = useContext(TreeFunctionContext);

  const configuration = state.configuration;

  const displayChildren = configuration.externalExpandNode
    ? state.expandedNodes && state.expandedNodes[node[configuration.key]]
    : node[configuration.displayChildren];

  let expandIcon = null;
  let collapsedIcon = null;
  let icon = null;
  if (
    !(
      node[configuration.expandIcon] ||
      node[configuration.collapsedIcon] ||
      node[configuration.icon]
    )
  ) {
    if (callbackContext.getIcons) {
      let icons = callbackContext.getIcons(node);
      expandIcon = icons.expandIcon;
      collapsedIcon = icons.collapsedIcon;
      icon = icons.icon;
    } else if (state.iconClass) {
      if (state.iconClass.operator) {
        expandIcon =
          state.iconClass.values[node[state.iconClass.operator]].expandedIcon;
        collapsedIcon =
          state.iconClass.values[node[state.iconClass.operator]].collapsedIcon;
        icon = state.iconClass.values[node[state.iconClass.operator]].icon;
      } else {
        expandIcon = state.iconClass.values.expandedIcon;
        collapsedIcon = state.iconClass.values.collapsedIcon;
        icon = state.iconClass.values.icon;
      }
    }
  }

  return (
    <>
      {node[configuration.expandIcon] && displayChildren ? (
        <i
          className={`${prefix}-tree-node-icon ${
            node[configuration.expandIcon]
          }`}
        />
      ) : expandIcon && displayChildren ? (
        cloneElement(expandIcon, {
          className: `${prefix}-tree-node-icon ${expandIcon.props.className}`
        })
      ) : null}

      {node[configuration.collapsedIcon] && !displayChildren ? (
        <i
          className={`${prefix}-tree-node-icon ${
            node[configuration.collapsedIcon]
          }`}
        />
      ) : collapsedIcon && !displayChildren ? (
        cloneElement(collapsedIcon, {
          className: `${prefix}-tree-node-icon ${collapsedIcon.props.className}`
        })
      ) : null}

      {node[configuration.icon] ? (
        <i className={`${prefix}-tree-node-icon ${node[configuration.icon]}`}>
          {' '}
        </i>
      ) : icon ? (
        cloneElement(icon, {
          className: `${prefix}-tree-node-icon ${icon.props.className}`
        })
      ) : null}
    </>
  );
};

NodeIcon.propTypes = {
  node: PropTypes.object
};

NodeIcon.defaultProps = {
  node: null
};

export default React.memo(NodeIcon);
