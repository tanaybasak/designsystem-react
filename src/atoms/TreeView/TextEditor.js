import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
import TextInput from '../TextInput';
import Overlay from '../Overlay';
const TextEditor = ({
  onClose,
  onTextUpdate,
  formStatus,
  errorMessage,
  ...restProps
}) => {
  const textEditorRef = useRef(null);
  const [width, setOverlayWidth] = useState('');
  const stopPropagation = e => {
    e.stopPropagation();
  };

  const updateTreenodeNameOnEnter = event => {
    event.stopPropagation();
    if (event.key === 'Enter') {
      onTextUpdate(event.currentTarget.value);
      event.preventDefault();
    } else if (event.key === 'Escape') {
      onClose();
      event.preventDefault();
    }
  };

  const updateTextNodeOnBlur = event => {
    onTextUpdate(event.currentTarget.value);
  };

  useEffect(() => {
    textEditorRef.current.firstElementChild.focus();
    textEditorRef.current.firstElementChild.select();
    setOverlayWidth(textEditorRef.current.firstElementChild.offsetWidth);
    return function cleanup() {};
  }, []);

  const [showOverlay, setOverlay] = useState(false);
  useEffect(() => {
    if (errorMessage) {
      setOverlay(true);
    }
  }, [errorMessage]);

  return (
    <div
      className={`${prefix}-form-group ${prefix}-text-container`}
      ref={textEditorRef}
    >
      <TextInput
        type="text"
        {...restProps}
        onBlur={updateTextNodeOnBlur}
        onKeyDown={updateTreenodeNameOnEnter}
        data-invalid={formStatus}
        onClick={stopPropagation}
      />
      {errorMessage ? (
        <Overlay
          showOverlay={showOverlay}
          targetElement={textEditorRef ? textEditorRef.current : null}
          attachElementToBody
          style={{ width: width }}
        >
          {errorMessage}
        </Overlay>
      ) : null}
    </div>
  );
};

TextEditor.propTypes = {
  onClose: PropTypes.func,
  onTextUpdate: PropTypes.func,
  formStatus: PropTypes.bool,
  errorMessage: PropTypes.any
};

TextEditor.defaultProps = {
  onClose: () => {},
  onTextUpdate: () => {},
  formStatus: false,
  errorMessage: null
};

export default TextEditor;
