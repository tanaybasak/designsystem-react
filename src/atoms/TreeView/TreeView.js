import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';
import TreeNode from './TreeNode';
import {
  updateTreeNode,
  updateNodePosition,
  deleteNode,
  copyNode
} from '../../util/treeUtil';

const TreeView = ({
  treeData,
  nodeSelected,
  expandedIcon,
  onChange,
  collapsedIcon,
  className,
  config,
  onToggle,
  type,
  onOverflowAction,
  iconClass,
  dragRules,
  isMoveNodeAllowed,
  isCopyAllowed,
  getOverFlowItems,
  onDeleteNode,
  onRenamingNode,
  onMoveNode,
  onCopyNode,
  onActionCompletes
}) => {
  let [treeInfo, updateTree] = useState(treeData);

  let [selectedNode, updateSelectedNode] = useState({});

  let [draggedNode, updateDraggedNode] = useState({});
  let [draggedNodeLevel, updateDraggedNodeLevel] = useState('');

  let [cutNode, updateCutNode] = useState({});
  let [cutNodeLevel, updateCutNodeLevel] = useState('');

  let [copiedNode, updateCopyNode] = useState({});

  useEffect(() => {
    updateTree(treeData);
  }, [treeData]);

  useEffect(() => {
    updateSelectedNode(nodeSelected);
  }, [nodeSelected]);

  const isMoveNodeAllowedMain = (x, y) => {
    return isMoveNodeAllowed ? isMoveNodeAllowed(x, y, treeInfo) : true;
  };

  const isCopyAllowedMain = (x, y) => {
    return isCopyAllowed ? isCopyAllowed(x, y, treeInfo) : true;
  };

  const onSelectNode = node => {
    updateSelectedNode(node);
    if (onChange) {
      onChange(node);
    }
  };

  const onToggleNode = event => {
    if (onToggle) {
      onToggle(event);
    }
  };

  let defaultConfig = {
    displayChildren: 'displayChildren',
    expandIcon: 'expandIcon',
    collapsedIcon: 'collapsedIcon',
    icon: 'icon',
    children: 'children',
    name: 'name',
    hasChildren: 'hasChildren',
    draggable: 'draggable'
  };

  const updateTreeState = (action, config) => {
    if (action === 'draggedNode') {
      updateDraggedNode(config.node);
      updateDraggedNodeLevel(config.level);
    } else if (action === 'cutNode') {
      updateCutNode(config.node);
      updateCopyNode(null);
      updateCutNodeLevel(config.level);
    } else if (action === 'copyNode') {
      updateCutNode(null);
      updateCopyNode(config.node);
      updateCutNodeLevel(null);
    }
  };

  const updateTreeDataBasedOnAction = async (action, config) => {
    if (action === 'delete') {
      let flag = onDeleteNode ? await onDeleteNode(config.node) : true;
      if (flag) {
        const updatedTree = deleteNode(treeInfo, config.level);
        updateTree(updatedTree);
        if (onActionCompletes) {
          onActionCompletes(action, updatedTree, config.node);
        }
      }
    } else if (action === 'copy-paste') {
      let flag = onCopyNode ? await onCopyNode(copiedNode, config.node) : true;
      if (flag) {
        const updatedTree = copyNode(treeInfo, config.level, config.copyNode);
        updateTree(updatedTree);
        if (onActionCompletes) {
          onActionCompletes('copy', updatedTree, config.copyNode, config.node);
        }
      }
    } else if (action === 'cut-paste') {
      let flag = onMoveNode ? await onMoveNode(cutNode, config.node) : true;
      if (flag) {
        const updatedTree = updateNodePosition(
          treeInfo,
          config.cutNodeLevel,
          config.level
        );
        updateTree(updatedTree);
        updateTreeState('cutNode', { node: null, level: '' });
        if (onActionCompletes) {
          onActionCompletes('cut', updatedTree, cutNode, config.node);
        }
      }
    } else if (action === 'toggle-node') {
      updateTree(updateTreeNode(treeInfo, config.node, config.level));
      if (onToggleNode) {
        onToggleNode(config.node);
      }
    } else if (action === 'node-update') {
      updateTree(updateTreeNode(treeInfo, config.node, config.level));
      if (onActionCompletes) {
        onActionCompletes('custom-action', treeInfo, config.node);
      }
    } else if (action === 'edit') {
      let flag = await onRenamingNode(config.node);
      if (flag) {
        const updatedTree = updateTreeNode(treeInfo, config.node, config.level);
        updateTree(updatedTree);
        if (onActionCompletes) {
          onActionCompletes(action, updatedTree, config.node);
        }
      }

      return flag;
    } else if (action === 'move-node') {
      let flag = onMoveNode ? await onMoveNode(draggedNode, config.node) : true;
      if (flag) {
        let dropNodeArray = config.dropNode.split('-');
        const dropNodeIndex = parseInt(dropNodeArray.splice(-1));
        if (dropNodeArray.length === 0) {
          dropNodeArray = null;
        } else {
          dropNodeArray = dropNodeArray.join('-');
        }
        const updatedTree = updateNodePosition(
          treeInfo,
          config.draggedNode,
          dropNodeArray,
          dropNodeIndex
        );
        updateTree(updatedTree);
        if (onActionCompletes) {
          onActionCompletes('drop', updatedTree, draggedNode, config.node);
        }
      }
    }
  };

  const configuration = { ...defaultConfig, ...config };
  const classnames = `${prefix}-tree ${className.trim()}`;
  return (
    <ul role="tree" className={classnames}>
      {treeInfo.map((node, index) => {
        return (
          <TreeNode
            node={node}
            key={`index-${index}`}
            expandedIcon={expandedIcon}
            collapsedIcon={collapsedIcon}
            iconClass={iconClass}
            dragRules={dragRules}
            isMoveNodeAllowed={isMoveNodeAllowedMain}
            isCopyAllowed={isCopyAllowedMain}
            draggedNode={draggedNode}
            draggedNodeLevel={draggedNodeLevel}
            level={index + ''}
            parentNode={null}
            onSelectNode={type === 'single' ? onSelectNode : null}
            selectedNode={type === 'single' ? selectedNode : null}
            configuration={configuration}
            onOverflowAction={onOverflowAction}
            copiedNode={copiedNode}
            cutNode={cutNode}
            getOverFlowItems={getOverFlowItems}
            cutNodeLevel={cutNodeLevel}
            updateTreeDataBasedOnAction={updateTreeDataBasedOnAction}
            updateTreeState={updateTreeState}
          />
        );
      })}
    </ul>
  );
};

TreeView.propTypes = {
  /** Tree Data */
  treeData: PropTypes.any,
  /** Used to pass icon classname for each node */
  iconClass: PropTypes.any,
  /** Used to specify draggable node */
  dragRules: PropTypes.any,
  /** Used to set selected node */
  nodeSelected: PropTypes.any,
  /** To Specify Expand Icon */
  expandedIcon: PropTypes.string,
  /** To Specify Collapsed Icon */
  collapsedIcon: PropTypes.string,
  /** Style class of the component */
  className: PropTypes.string,
  /** Callback function on selecting overflow menu item */
  onOverflowAction: PropTypes.func,
  /** Callback function used for specifying rules on drag and drop or cut and paste */
  isMoveNodeAllowed: PropTypes.func,
  /** Callback function used for specifying rules on copy and paste */
  isCopyAllowed: PropTypes.func,
  /** Callback function on selecting tree node */
  onChange: PropTypes.func,
  /** Callback function on expanding/collapsing tree node */
  onToggle: PropTypes.func,
  /** Callback function on deleting tree node from overflow menu */
  onDeleteNode: PropTypes.func,
  /** Callback function for setting overflow menu actions
   *  eg:
   * [
   * {name: 'Rename',action: 'edit'}
   * {name: 'Cut',action: 'cut'}
   * {name: 'Copy',action: 'copy'}
   * {name: 'Delete',action: 'delete'}
   * ]
   *
   */
  getOverFlowItems: PropTypes.func,
  /** Callback function on renaming tree node from overflow menu */
  onRenamingNode: PropTypes.func,
  /** Configuration Object for updating propery name in tree data
 {
  displayChildren: 'displayChildren',
  expandIcon: 'expandIcon',
  collapsedIcon: 'collapsedIcon',
  icon: 'icon',
  children: 'children',
  name: 'name',
  hasChildren: 'hasChildren'
}
*/
  config: PropTypes.any,
  /** Type of Treeview
   * default : Component without any node selection
   * single : Component with node selection
   */
  type: PropTypes.oneOf(['default', 'single']),
  /** Callback function on moving node  */
  onMoveNode: PropTypes.func,
  /** Callback function on pasting node  */
  onCopyNode: PropTypes.func,
  /** Callback function after completing the overflow action  */
  onActionCompletes: PropTypes.func
};

TreeView.defaultProps = {
  treeData: [],
  iconClass: null,
  dragRules: null,
  onChange: null,
  onToggle: null,
  isMoveNodeAllowed: null,
  isCopyAllowed: null,
  onRenamingNode: null,
  expandedIcon: 'caret caret-down',
  collapsedIcon: 'caret',
  className: '',
  type: 'default',
  config: {},
  onOverflowAction: null,
  getOverFlowItems: null,
  nodeSelected: null,
  onMoveNode: null,
  onCopyNode: null,
  onDeleteNode: null,
  onActionCompletes: null
};

export default TreeView;
