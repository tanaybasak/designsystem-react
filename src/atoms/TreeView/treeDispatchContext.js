import React from 'react';
const TreeDispatchContext = React.createContext({
  dispatch: () => {}
});
export default TreeDispatchContext;
