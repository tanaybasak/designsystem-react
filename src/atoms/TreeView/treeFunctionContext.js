import React from 'react';
const TreeFunctionContext = React.createContext({
  onChange: null,
  onToggle: null,
  onDeleteNode: null,
  onCopyNode: null,
  onMoveNode: null,
  onActionCompletes: null,
  onOverflowAction: null,
  onRenamingNode: null,
  isCopyAllowed: null,
  isMoveNodeAllowed: null,
  isDropAllowed: null,
  getOverFlowItems: null,
  customTemplate: null,
  customActionTemplate: null,
  customNodeTemplate: null,
  getIcons: null,

  onDragStart: null,
  onDragOver: null,
  onDragEnter: null,
  onDragLeave: null,
  onDrop: null,
  onDragEnd: null,
  onDoubleClick: null,
  onKeyDown: null,
  onClick: null,
  isDraggable: null
});
export default TreeFunctionContext;
