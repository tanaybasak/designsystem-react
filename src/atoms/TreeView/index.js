import TreeView from './TreeView';
import TreeNodeTemplate from './TreeNodeTemplate';

export { TreeNodeTemplate, TreeView };
