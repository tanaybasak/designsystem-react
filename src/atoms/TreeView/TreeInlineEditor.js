import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import TextEditor from './TextEditor';
import { updateTreeNode } from '../../util/treeUtil';
import TreeDispatchContext from './treeDispatchContext';
import TreeStateContext from './treeStateContext';
import TreeFunctionContext from './treeFunctionContext';
const TreeInlineEditor = ({ content, node, level }) => {
  const [formStatus, updateFormStaus] = useState(false);
  const [errorMessage, updateErrorMessage] = useState(null);

  const dispatch = useContext(TreeDispatchContext);
  const state = useContext(TreeStateContext);
  const callbackContext = useContext(TreeFunctionContext);

  const configuration = state.configuration;

  const updateNodeText = value => {
    updateTreeNodeName(value);
  };

  const updateTreeNodeName = async value => {
    let nodeTemp = { ...node };
    if (nodeTemp[configuration.name] !== value) {
      nodeTemp[configuration.name] = value;

      let flag = await callbackContext.onRenamingNode(nodeTemp);
      if (flag[0]) {
        const updatedTree = updateTreeNode(
          state.treeInfo,
          nodeTemp,
          level,
          configuration
        );
        dispatch({ type: 'SET_TREE_DATA', data: updatedTree });
        if (callbackContext.onActionCompletes) {
          callbackContext.onActionCompletes('edit', updatedTree, nodeTemp);
        }

        updateFormStaus(false);
        updateErrorMessage('');
        dispatch({ type: 'SET_RENAME_NODE_ID', data: null });
      } else {
        updateFormStaus(true);
        updateErrorMessage(flag[1]);
      }
    } else {
      updateFormStaus(false);
      dispatch({ type: 'SET_RENAME_NODE_ID', data: null });
    }
  };

  const closeTextField = () => {
    dispatch({ type: 'SET_RENAME_NODE_ID', data: null });
    updateFormStaus(false);
  };

  return (
    <TextEditor
      value={content}
      onTextUpdate={updateNodeText}
      formStatus={formStatus}
      errorMessage={errorMessage}
      onClose={closeTextField}
    />
  );
};

TreeInlineEditor.propTypes = {
  content: PropTypes.any,
  node: PropTypes.object,
  level: PropTypes.string
};

TreeInlineEditor.defaultProps = {
  content: null,
  node: null,
  level: null
};

export default TreeInlineEditor;
