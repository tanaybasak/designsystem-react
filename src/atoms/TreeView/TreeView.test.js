import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TreeView from './TreeView';

const treeData = [
  {
    name: 'Section 1',
    displayChildren: false,
    collapsedIcon: 'section1expand',
    expandIcon: 'section1collapsed',
    children: [
      {
        name: 'Sub Section 1.1',
        icon: 'subsection1.1icon',
        displayChildren: false
      },
      {
        name: 'Sub Section 1.2',
        icon: 'subsection1.2icon',
        displayChildren: false,
        children: []
      }
    ]
  },
  {
    name: 'Section 2',
    displayChildren: false,
    collapsedIcon: 'section1expand',
    expandIcon: 'section1collapsed',
    children: []
  }
];

it('TreeView renders correctly', () => {
  const tree = renderer
    .create(
      <TreeView treeData={treeData} onChange={() => {}} onToggle={() => {}} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Toggle Tree node using click', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} onToggle={mockCallBack} />
  );
  const firstTreeNode = treeData[0];
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();

  wrapper
    .find('.tree-node')
    .at(0)
    .simulate('click');
  expect(wrapper.find(`.tree-nested`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);
});

it('Select Tree Node', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} type="single" onChange={mockCallBack} />
  );
  expect(wrapper.find(`.highlight`).exists()).toBeFalsy();
  wrapper
    .find('.tree-node')
    .at(0)
    .find('span')
    .simulate('click');
  expect(wrapper.find(`.highlight`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);
});

it('Select Tree on Key enter', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} type="single" onChange={mockCallBack} />
  );
  const firstTreeNode = treeData[0];
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();
  wrapper
    .find('.tree-node')
    .at(0)
    .simulate('focus');
  wrapper
    .find('.tree-node')
    .at(0)
    .simulate('keydown', { keyCode: 13 });
  expect(wrapper.find(`.highlight`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);
});

it('Toggle Tree Node on right Arrow and left arrow', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} onToggle={mockCallBack} />
  );
  const firstTreeNode = treeData[0];
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();

  const treeNode = wrapper.find('.tree-node').at(0);
  treeNode.simulate('focus');
  treeNode.simulate('keydown', { keyCode: 39 });
  expect(wrapper.find(`.tree-nested`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);

  treeNode.simulate('focus');
  treeNode.simulate('keydown', { keyCode: 37 });
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();
});

it('Navigate Item using Arrow Down', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} onToggle={mockCallBack} />
  );
  const firstTreeNode = treeData[0];
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();

  const treeNode = wrapper.find('.tree-node').at(0);
  treeNode.simulate('focus');
  treeNode.simulate('keydown', { keyCode: 39 });
  expect(wrapper.find(`.tree-nested`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);
  treeNode.simulate('keydown', { keyCode: 40 });
  expect(document.activeElement.textContent.trim()).toEqual(
    treeData[0].children[0].name
  );
  let currentActiveNode = wrapper.find('.tree-node').at(1);
  currentActiveNode.simulate('keydown', { keyCode: 40 });
  expect(document.activeElement.textContent.trim()).toEqual(
    treeData[0].children[1].name
  );

  currentActiveNode = wrapper.find('.tree-node').at(2);
  currentActiveNode.simulate('keydown', { keyCode: 40 });
  expect(document.activeElement.textContent.trim()).toEqual(treeData[1].name);
});

it('Navigate Item using Arrow Up', () => {
  const mockCallBack = jest.fn();
  const wrapper = mount(
    <TreeView treeData={treeData} onToggle={mockCallBack} />
  );
  const firstTreeNode = treeData[0];
  expect(wrapper.find(`.tree-nested`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeFalsy();

  const firstNode = wrapper.find('.tree-node').at(0);
  const secondNode = wrapper.find('.tree-node').at(1);
  firstNode.simulate('focus');
  firstNode.simulate('keydown', { keyCode: 39 });
  expect(wrapper.find(`.tree-nested`).exists()).toBeTruthy();
  expect(wrapper.find(`.${firstTreeNode.collapsedIcon}`).exists()).toBeFalsy();
  expect(wrapper.find(`.${firstTreeNode.expandIcon}`).exists()).toBeTruthy();
  expect(mockCallBack.mock.calls.length).toBe(1);

  secondNode.simulate('focus');
  expect(document.activeElement.textContent.trim()).toEqual(treeData[1].name);
  secondNode.simulate('keydown', { keyCode: 38 });

  expect(document.activeElement.textContent.trim()).toEqual(
    treeData[0].children[1].name
  );

  wrapper
    .find('.tree-node')
    .at(2)
    .simulate('keydown', { keyCode: 38 });
  expect(document.activeElement.textContent.trim()).toEqual(
    treeData[0].children[0].name
  );

  wrapper
    .find('.tree-node')
    .at(1)
    .simulate('keydown', { keyCode: 38 });
  expect(document.activeElement.textContent.trim()).toEqual(treeData[0].name);
});
