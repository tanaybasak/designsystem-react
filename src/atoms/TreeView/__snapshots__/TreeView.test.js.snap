// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView renders correctly 1`] = `
<ul
  className="hcl-tree "
  role="tree"
>
  <li
    aria-expanded={false}
    className="tree-item"
    role="treeitem"
  >
    <div
      className="tree-node "
      level="0"
      onClick={[Function]}
      onDragEnd={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <i
        className="toggle-icon caret"
      />
      <div
        className="treenode-action-wrapper"
      >
        <div
          className="treenode-content-wrapper"
        >
          <i
            className="section1expand"
          />
          <span
            className="tree-text-node "
            draggable={false}
            onClick={[Function]}
            onDragStart={[Function]}
            title="Section 1"
          >
            <div
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              Section 1
            </div>
          </span>
        </div>
      </div>
    </div>
  </li>
  <li
    aria-expanded={null}
    className="tree-item"
    role="treeitem"
  >
    <div
      className="tree-node no-children-exist "
      level="1"
      onDragEnd={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <div
        className="treenode-action-wrapper"
      >
        <div
          className="treenode-content-wrapper"
        >
          <i
            className="section2expand"
          />
          <span
            className="tree-text-node "
            draggable={false}
            onClick={[Function]}
            onDragStart={[Function]}
            title="Section 2"
          >
            <div
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              Section 2
            </div>
          </span>
        </div>
      </div>
    </div>
  </li>
</ul>
`;
